osu.Game/Overlays/Chat/ChannelScrollContainer.cs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Graphics;
+using osu.Game.Graphics.Containers;
+
+namespace osu.Game.Overlays.Chat
+{
+    /// <summary>
+    /// An <see cref="OsuScrollContainer"/> with functionality to automatically scroll whenever the maximum scrollable distance increases.
+    /// </summary>
+    public class ChannelScrollContainer : OsuScrollContainer
+    {
+        /// <summary>
+        /// The chat will be automatically scrolled to end if and only if
+        /// the distance between the current scroll position and the end of the scroll
+        /// is less than this value.
+        /// </summary>
+        private const float auto_scroll_leniency = 10f;
+
+        /// <summary>
+        /// Whether to keep this container scrolled to end on new content.
+        /// </summary>
+        /// <remarks>
+        /// This is specifically controlled by whether the latest scroll operation made the container scrolled to end.
+        /// </remarks>
+        private bool trackNewContent = true;
+
+        protected override void UpdateAfterChildren()
+        {
+            base.UpdateAfterChildren();
+
+            if (trackNewContent && !IsScrolledToEnd())
+                ScrollToEnd();
+        }
+
+        private void updateTrackState() => trackNewContent = IsScrolledToEnd(auto_scroll_leniency);
+
+        // todo: we may eventually want this encapsulated in a "OnScrollChange" event handler method provided by ScrollContainer.
+        // important to note that this intentionally doesn't consider OffsetScrollPosition, but could make it do so with side changes.
```

Comment:
So basically, because `OffsetScrollPosition` isn't handled, deleting messages won't update the tracked state. But this shouldn't result in a bad result in most cases.